src/mds/Locker.cc

```diff
@@ -5485,3 +5501,17 @@ void Locker::handle_file_lock(ScatterLock *lock, const cref_t<MLock> &m)
     ceph_abort();
   }  
 }
+
+void Locker::lock_stablized(SimpleLock* lock, uint64_t mask, int r, MDSContext::vec* finishers) {
+  if (lock->get_type() == CEPH_LOCK_INEST && lock->is_dirty()) {
+    dout(20) << __func__ << " lock " << *lock << " on " << *lock->get_parent() 
+    << " pushed to the back of updated_scatterlocks due to quickflush" << dendl;
+    mark_updated_scatterlock(static_cast<ScatterLock*>(lock));
+  }
+
+  if (finishers)
+    lock->take_waiting(mask, *finishers);
+  else
+    lock->finish_waiters(mask, r);
+}
```

Comment:
for the first case in your commit message. the nestlock should already be nudging, I don't see why need to add it to updated_scatterlocks.

for the second case, I think it's easy to make scatter_nudge to re-queue lock to updated_scatterlocks

Reply:
Hi, I thought the first case may not always be triggered by a scatter_nudge. it may be triggered by a wrlock_start or something like that, right? So I that it would be safe to uniformly requeue all those nestlocks that just turned stable to updated_scatterlocks.

Am I right?

Reply:
yes, it can happen at xxx_start().    but nestlock's is in LOCK_MIX_FOO state, it will gather dirty rstats.

Reply:
I think I understand that, but these two are only the known cases that lead the lost of dirty rstat, there may be others as well. I added this to make sure that dirty rstat are always taken care of, i don't see why it has to be removed

Reply:
for the second case, there doesn't be much difference between putting the lock back to updated_scatterlocks, letting the waiter context to scatter_nudge it and scatter_nudging it, letting the waiter context to.check on the lock and wait for it.

besides, these are only cases that i encountered when doing the test, there could be other cases. putting the locks back to updated_scatterlocks can prevent all those cases from happenning once and for all, and even if there isn't other such cases, doing this wouldn't bring anything bad. Right?

Reply:
oh, i might have misunderstood you. do you mean that i should simply change the "mark_updated_scatterlock" to a "scatter_nudge"?

Reply:
I'm not fan of the lock_stablized approach. I prefer
1. add lock to updated_scatterlocks when it become dirty
2. if scatter tick failed to nudge a lock, re-queue it to updated_scatterlocks or another list

Reply:
yes, i'm already doing that, my approach is like this:
1. add lock to updated_scatterlocks when it become dirty
2. if scatter tick fails to nudge a lock, queue it to nudging_nestlocks

but, there are circumstances that when a lock finished a nudge, it is still dirty and not in the updated_scatterlocks, so we have to put it back. If we don't add a method like lock_stablized, we have add a mark_updated_scatterlocks anywhere there is a "finish_waiters(WAIT_STABLE)" or "take_waiting(WAIT_STABLE)". this would be redundant code, right? Why don't be process the finish_waiter(WAIT_STABLE) at a uniform place? Am I right?

Reply:
I think I know what's the problem here.  Although the scatter_nudge's return is a bool value, it's not an indicator whether the lock is nudged immediately or nudge-able.

Actually, even if the lock is nudged immediately, the scatter_nudge will still end up waiting for the lock to be stable. it seems there is currently no way to judge, even in the scatter_nudge method, whether the waiting is caused by an immediate nudge or an lock's scatter-gather process.

The reason why i changed to scatter_nudge's return type to bool is that, at the time, I can't be sure whether the waiter i passed to scatter_nudge will finally be added to some waiting list, if not, i have to delete it to prevent a memory leak, so i changed its return type to bool just to indicate whether the waiter is on some list.

i'm sorry for this miss understanding. i think i can improve that part of the code:). really sorry.

Reply:
the scatter_tick is also calling nudge_updated_scatterlocks now

Reply:
@ukernel Hi:-)

I think I can change the scatter_mix, simple_lock and simple_sync to return whether the lock is nudged or undergoing a scatter-gather process, so that the scatter_nudge can indicate whether the lock is nudged immediately, just like what I did in the previous submit. 

Then I can make "nudge_updated_scatterlock" to requeue the not-nudged-immediately nestlocks to updated_scatterlocks to prevent dirty rstat nestlock leak. I think this is what you are suggesting, right? Thanks:-)

Reply:
@ukernel Or, we may not need to change scatter_mix, simple_lock and simple_sync; we can just make scatter_nudge treat an invocation of scatter_mix, simple_lock or simple_sync as a indication of a successful nudge, as scatter_gather can be seen as part of a immediate nudge. Am I right? Thanks:-)

Reply:
@ukernel But I think, although we don't have to put the dirty nestlocks back to updated_scatterlocks when they turn stable, we still have to move them out of the "nudging_nestlocks" queue after they turn stable, so, once agian, a method like lock_stablized seems inevitable, is this right?

Reply:
@ukernel Or, I think maybe I can keep locks in updated_scatterlocks until it's not dirty. I mean, instead of poping every lock that needs to be nudged, we keep them in the queue even if they are nudged immediately and let later run of nudge_updated_scatterlocks() to seletively remove not-dirty locks. How do you like this way:-) Thanks:-)

Reply:
@ukernel Or, I think maybe we can change the nudging_nestlocks to flushing_nestlocks. And instead of poping every lock out of the updated_scatterlocks queue, we make nudge_updated_scatterlocks() not modify updated_scatterlocks queue, and let the Scatterlock::start_flush() to move a lock from updated_scatterlocks to flushing_nestlocks if that lock's parent doesn't have children with dirty rstat. With this approach, no dirty lock would be lost, since it's only when they "start_flush" and doesn't have any children with dirty rstat does they get to be remove from updated_scatterlocks(), and we can avoid the lock_stablized likens. Am I right about this? Thanks:-)

Reply:
@ukernel I think maybe we can remove the nudging_nestlocks and only rely on the updated_scatterlocks. The approach is like this: instead of poping every lock from the updated_scatterlocks, we make nudge_updated_scatterlock not change the updated_scatterlocks and let the ScatterLock::start_flush to remove the lock from updated_scatterlocks if the lock doesn't have any children with dirty rstat, which is exactly what "not dirty" means. Meanwhile, we make sure every nudge eventually lead to a "start_flush". With this approach, we can avoid the lock_stablized likes and make sure no dirty rstat is lost. Do you think this is feasible? Thanks:-)

Reply:
> for the first case in your commit message. the nestlock should already be nudging, I don't see why need to add it to updated_scatterlocks.
> 
> for the second case, I think it's easy to make scatter_nudge to re-queue lock to updated_scatterlocks

Finally, I got your point:-)

Reply:
@ukernel Sorry that I didn't catch up with you for so long:-)